File: app/src/main/java/com/example/sudoku_solver/MainActivity.java
package com.example.sudoku_solver;

import androidx.annotation.MainThread;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;

import android.Manifest;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.graphics.Bitmap;
import android.net.Uri;
import android.os.Bundle;
import android.provider.MediaStore;
import android.util.Log;
import android.view.SurfaceView;
import android.view.View;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.Toast;

import com.google.android.gms.tasks.OnFailureListener;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.android.gms.tasks.Task;
import com.google.mlkit.vision.common.InputImage;
import com.google.mlkit.vision.text.Text;
import com.google.mlkit.vision.text.TextRecognition;
import com.google.mlkit.vision.text.TextRecognizer;
import com.google.mlkit.vision.text.latin.TextRecognizerOptions;

import org.opencv.android.BaseLoaderCallback;
import org.opencv.android.CameraBridgeViewBase;
import org.opencv.android.JavaCameraView;
import org.opencv.android.OpenCVLoader;
import org.opencv.android.Utils;
import org.opencv.core.CvType;
import org.opencv.core.Mat;
import org.opencv.core.MatOfPoint;
import org.opencv.core.MatOfPoint2f;
import org.opencv.core.Point;
import org.opencv.core.Rect;
import org.opencv.core.Scalar;
import org.opencv.core.Size;
import org.opencv.imgproc.Imgproc;

import java.io.FileNotFoundException;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

public class MainActivity extends AppCompatActivity implements CameraBridgeViewBase.CvCameraViewListener2 {

    private Mat mRGBA, mRGBAT;
    private Rect rect = null;
    private JavaCameraView javaCameraView;
    private Button camButton;
    private Button flashButton;
    private Button keyboardButton;
    private boolean ok = true;
    private boolean flash = false;
    public static int[][] board;
    private int minNumbersOnBoard = 7;
    private int curNumbersOnBoard = 0;
    private int SOLVE_FROM_IMAGE = 0;
    private int SOLVE_FROM_MANUAL = 1;
    private final TextRecognizer textRecognizer = TextRecognition.getClient(TextRecognizerOptions.DEFAULT_OPTIONS);

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        javaCameraView = (JavaCameraView) findViewById(R.id.camera_surface);
        javaCameraView.setVisibility(SurfaceView.VISIBLE);
        javaCameraView.setCvCameraViewListener(MainActivity.this);

        camButton = (Button) findViewById(R.id.camBtn);
        flashButton = (Button) findViewById(R.id.flashBtn);
        keyboardButton = (Button) findViewById(R.id.kBtn);

        board = new int[9][9];

        if (ContextCompat.checkSelfPermission(this, Manifest.permission.CAMERA) == PackageManager.PERMISSION_DENIED) {
            ActivityCompat.requestPermissions(MainActivity.this, new String[]{Manifest.permission.CAMERA}, 112);
        }

        keyboardButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                javaCameraView.disableView();

                Intent i = new Intent(MainActivity.this, SolverActivity.class);
                i.putExtra("type", SOLVE_FROM_MANUAL);
                startActivity(i);
            }
        });

        flashButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                if (!flash) {
                    javaCameraView.turnOnFlashlight();
                    flashButton.setBackgroundResource(R.drawable.flash_on);
                    flash = true;
                } else {
                    javaCameraView.turnOffFlashLight();
                    flashButton.setBackgroundResource(R.drawable.flash_off);
                    flash = false;
                }
            }
        });

        camButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                ok = false;
                Mat mat = processImage(mRGBAT, rect);
                recognizeTextFromImage(mat);
            }
        });
    }

    public Mat processImage(Mat mat, Rect rect) {
        Mat binaryImage = new Mat();
        Mat lines = new Mat();
        List<MatOfPoint> contours = new ArrayList<MatOfPoint>();

        // transform image to binary image
        double thresh = Imgproc.threshold(mat, binaryImage, 0, 255, Imgproc.THRESH_BINARY_INV + Imgproc.THRESH_OTSU);
        //Imgproc.medianBlur(mat, mat, 5);
        //Imgproc.adaptiveThreshold(mat, binaryImage, 255, Imgproc.ADAPTIVE_THRESH_GAUSSIAN_C, Imgproc.THRESH_BINARY_INV, 11, 2);


        // remove all horizontal lines
        Mat hkernel = Imgproc.getStructuringElement(Imgproc.MORPH_RECT, new Size(50, 1));
        Imgproc.morphologyEx(binaryImage, lines, Imgproc.MORPH_OPEN, hkernel);
        Imgproc.findContours(lines, contours, new Mat(), Imgproc.RETR_EXTERNAL, Imgproc.CHAIN_APPROX_SIMPLE);
        Imgproc.drawContours(binaryImage, contours, -1, new Scalar(0, 0, 0), 10);

        // remove all vertical lines
        Mat vkernel = Imgproc.getStructuringElement(Imgproc.MORPH_RECT, new Size(1, 40));
        Imgproc.morphologyEx(binaryImage, lines, Imgproc.MORPH_OPEN, vkernel);
        Imgproc.findContours(lines, contours, new Mat(), Imgproc.RETR_EXTERNAL, Imgproc.CHAIN_APPROX_SIMPLE);
        Imgproc.drawContours(binaryImage, contours, -1, new Scalar(0, 0, 0), 10);

        //crop image
        Mat cropped = new Mat(binaryImage, rect);
        Imgproc.resize(cropped, cropped, mat.size());

        return cropped;
    }

    public Rect getSudokuBoundingRectangle(Mat mRGBA) {

        // remove noise from image, transform to grayscale
        Mat mRGBAG = new Mat();
        Mat canny = new Mat();
        Imgproc.GaussianBlur(mRGBA, mRGBAG, new Size(3, 3), 0, 0);
        Imgproc.cvtColor(mRGBAG, mRGBAG, Imgproc.COLOR_RGB2GRAY);
        Imgproc.Canny(mRGBAG, canny, 200, 255);

        // get all the contours
        List<MatOfPoint> contours = new ArrayList<MatOfPoint>();
        Imgproc.findContours(canny, contours, new Mat(), Imgproc.RETR_TREE, Imgproc.CHAIN_APPROX_SIMPLE);

        // if there are contours, get the biggest one
        if (contours.size() > 0) {
            double maxA = 0;
            int imax = 0;
            for (int i = 0; i < contours.size(); i++) {
                Mat contour = contours.get(i);
                if (Imgproc.contourArea(contour) > maxA) {
                    maxA = Imgproc.contourArea(contour);
                    imax = i;
                }
            }

            Imgproc.drawContours(mRGBA, contours, imax, new Scalar(0, 255, 0), 1, Imgproc.LINE_AA);

            // get the bounding box
            MatOfPoint2f c2f = new MatOfPoint2f(contours.get(imax).toArray());
            MatOfPoint2f approx = new MatOfPoint2f();
            double perimeter = Imgproc.arcLength(c2f, true);
            Imgproc.approxPolyDP(c2f, approx, 0.01 * perimeter, true);

            MatOfPoint points = new MatOfPoint(approx.toArray());
            Rect rect = Imgproc.boundingRect(points);

            return rect;
        } else {
            return null;
        }
    }

    public void recognizeTextFromImage(Mat mRGBA) {
        Bitmap bitmapImage = Bitmap.createBitmap(mRGBA.cols(), mRGBA.rows(), Bitmap.Config.ARGB_8888);
        Utils.matToBitmap(mRGBA, bitmapImage);
        InputImage inImg = InputImage.fromBitmap(bitmapImage, 90);
        Task<Text> result = textRecognizer.process(inImg)
                .addOnSuccessListener(new OnSuccessListener<Text>() {
                    @Override
                    public void onSuccess(Text text) {
                        Log.e("MLKIT", "Task success");
                        for (Text.TextBlock block : text.getTextBlocks()) {
                            for (Text.Line line : block.getLines()) {
                                for (Text.Element element : line.getElements()) {
                                    android.graphics.Rect elRect = element.getBoundingBox();

                                    int approxWidth = mRGBA.width() / 9;
                                    int approxHeight = mRGBA.height() / 9;

                                    int j = elRect.centerX() / approxWidth;
                                    int i = elRect.centerY() / approxHeight;

                                    char c = transformPossibleNumberFromChar(element.getText().charAt(0));
                                    int nr = c - '0';
                                    curNumbersOnBoard++;
                                    board[i][j] = nr > 0 && nr < 10 ? nr : 0;
                                }
                            }
                        }

                        if (check()) {
                            Intent i = new Intent(MainActivity.this, SolverActivity.class);
                            i.putExtra("type", SOLVE_FROM_IMAGE);
                            i.putExtra("board", board);
                            startActivity(i);
                        } else {
                            Toast toast = Toast.makeText(getApplicationContext(), "PLEASE RETAKE PHOTO", Toast.LENGTH_SHORT);
                            toast.show();
                            ok = true;
                        }

                    }
                }).addOnFailureListener(new OnFailureListener() {
                    @Override
                    public void onFailure(@NonNull Exception e) {
                        Log.e("MLKIT", "Task fail");
                    }
                });
    }

    private boolean check() {
        if (curNumbersOnBoard < minNumbersOnBoard || (rect.width < mRGBA.width() / 3 && rect.height < mRGBA.height() / 3)) {
            return false;
        }
        return true;
    }

    private char transformPossibleNumberFromChar(char c) {
        switch (c) {
            case 'A':
                c = '4';
                break;
            case '|':
                c = '1';
                break;
        }
        return c;
    }

    private BaseLoaderCallback baseLoaderCallback = new BaseLoaderCallback(this) {
        @Override
        public void onManagerConnected(int status) {
            switch (status) {
                case BaseLoaderCallback.SUCCESS:
                    javaCameraView.enableView();
                    break;
                default:
                    super.onManagerConnected(status);
                    break;
            }
        }
    };

    @Override
    public Mat onCameraFrame(CameraBridgeViewBase.CvCameraViewFrame inputFrame) {
        if (ok) {
            mRGBA = inputFrame.rgba();
            mRGBAT = inputFrame.gray();
            rect = getSudokuBoundingRectangle(mRGBA);
        }
        return mRGBA;
    }

    @Override
    public void onCameraViewStarted(int width, int height) {
        mRGBA = new Mat(height, width, CvType.CV_8UC4);
        mRGBAT = new Mat(height, width, CvType.CV_8UC1);
    }

    @Override
    protected void onResume() {
        super.onResume();
        if (OpenCVLoader.initDebug()) {
            baseLoaderCallback.onManagerConnected(BaseLoaderCallback.SUCCESS);
        } else {
            OpenCVLoader.initAsync(OpenCVLoader.OPENCV_VERSION, this, baseLoaderCallback);
        }
    }

    @Override
    protected void onPause() {
        super.onPause();
        if (javaCameraView != null) {
            javaCameraView.disableView();
        }
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        if (javaCameraView != null) {
            javaCameraView.disableView();
        }
    }

    @Override
    public void onCameraViewStopped() {
        mRGBA.release();
    }
}

File: app/src/main/java/com/example/sudokusolver2/MainActivity.kt
package com.example.sudokusolver2

import android.Manifest
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.graphics.Bitmap
import android.hardware.camera2.CameraManager
import android.os.Bundle
import android.util.Log
import android.view.SurfaceView
import android.widget.Button
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import com.example.sudokusolver2.databinding.ActivityMainBinding
import com.example.sudokusolver2.imageproc.OCR
import com.example.sudokusolver2.imageproc.OpenCV
import com.google.mlkit.vision.common.InputImage
import kotlinx.coroutines.runBlocking
import org.opencv.android.BaseLoaderCallback
import org.opencv.android.CameraBridgeViewBase
import org.opencv.android.JavaCameraView
import org.opencv.android.LoaderCallbackInterface
import org.opencv.android.OpenCVLoader
import org.opencv.android.Utils
import org.opencv.core.CvType
import org.opencv.core.Mat

class MainActivity : AppCompatActivity(), CameraBridgeViewBase.CvCameraViewListener2 {

    companion object {
        private const val CAMERA_PERMISSION_CODE = 101
        private const val TAG = "SudokuSolver2"

        private const val STATUS_NOTHING = -1
        private const val STATUS_PREVIEWING = 1
        private const val STATUS_PROCESS_IMAGE = 2
    }

    private lateinit var javaCameraView: JavaCameraView
    private lateinit var mRGBA: Mat

    private var STATUS = STATUS_PREVIEWING
    private var FLASH_STATE = false

    private lateinit var viewBinding: ActivityMainBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        viewBinding = ActivityMainBinding.inflate(layoutInflater, null, false)
        setContentView(viewBinding.root)

        // Activate the camera only if we have the necessary permission, otherwise require the permission
        if (ContextCompat.checkSelfPermission(this, Manifest.permission.CAMERA) == PackageManager.PERMISSION_GRANTED) {
            activateCamera()
        } else {
            requestPermissions(arrayOf(Manifest.permission.CAMERA), CAMERA_PERMISSION_CODE)
        }

        findViewById<Button>(R.id.camButton).setOnClickListener {
            STATUS = STATUS_PROCESS_IMAGE
        }

        Solver.init(this)

        /**
         * Turn flashlight on / off
         * https://answers.opencv.org/question/131404/opencv-android-implement-flashlight/
         */
        findViewById<Button>(R.id.flashLight).setOnClickListener {
            FLASH_STATE = !FLASH_STATE

            if (FLASH_STATE) {
                javaCameraView.turnOnFlashlight()
                it.setBackgroundResource(R.drawable.baseline_flash_on_24)
            } else {
                javaCameraView.turnOffFlashLight()
                it.setBackgroundResource(R.drawable.baseline_flash_off_24)
            }
        }

        findViewById<Button>(R.id.keyboard).setOnClickListener {
            startActivity(Intent(baseContext, SolverActivity::class.java).apply {
                addFlags(Intent.FLAG_ACTIVITY_NO_ANIMATION)
            })
        }
    }

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<out String>,
        grantResults: IntArray
    ) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)

        if (requestCode == CAMERA_PERMISSION_CODE) {
            if (!grantResults.contains(PackageManager.PERMISSION_DENIED)) {
                activateCamera()
            } else {
                Toast.makeText(
                    applicationContext,
                    "Camera permission is required",
                    Toast.LENGTH_LONG
                ).show()
            }
        }
    }

    // Setup for the JavaCameraView
    private fun activateCamera() {
        javaCameraView = findViewById(R.id.javaCamView)

        javaCameraView.setCameraPermissionGranted()
        javaCameraView.setCameraIndex(CameraBridgeViewBase.CAMERA_ID_BACK)
        javaCameraView.visibility = SurfaceView.VISIBLE
        javaCameraView.setCvCameraViewListener(this)
        javaCameraView.enableView()
    }

    private val mBaseLoaderCallback: BaseLoaderCallback = object : BaseLoaderCallback(this) {
        override fun onManagerConnected(status: Int) {
            when (status) {
                LoaderCallbackInterface.SUCCESS -> {
                    Log.d(TAG, "OpenCV loaded!")
                    activateCamera()
                }

                else -> super.onManagerConnected(status)
            }
        }
    }

    // Initialize (Load) the OpenCV library
    override fun onResume() {
        super.onResume()
        if (OpenCVLoader.initDebug()) {
            mBaseLoaderCallback.onManagerConnected(BaseLoaderCallback.SUCCESS)
        } else {
            OpenCVLoader.initAsync(OpenCVLoader.OPENCV_VERSION, this, mBaseLoaderCallback)
        }
    }

    override fun onCameraViewStarted(width: Int, height: Int) {
        mRGBA = Mat(width, height, CvType.CV_8UC4)
    }

    override fun onCameraViewStopped() {
        mRGBA.release()
    }

    override fun onCameraFrame(inputFrame: CameraBridgeViewBase.CvCameraViewFrame?): Mat? {
        if (STATUS == STATUS_PREVIEWING) {
            mRGBA = inputFrame!!.rgba()
            OpenCV.getSudokuBoundingRectangle(mRGBA, true)

            return mRGBA
        } else if (STATUS == STATUS_PROCESS_IMAGE) {
            // If the camera button was pressed prepare the image for processing

            // Step 1: Get the bounding rectangle's corner points
            val cntPts = OpenCV.getSudokuBoundingRectangle(mRGBA, false)

            // If the OpenCV.getSudokuBoundingRectangle returns null don't continue and return to previewing
            if (cntPts == null) {
                runOnUiThread {
                    Toast.makeText(
                        applicationContext,
                        "Couldn't detect any sudoku board. Try again!",
                        Toast.LENGTH_SHORT
                    ).show()
                }
            } else {
                // Step 2: Order the points clockwise
                val sortedPoints = OpenCV.orderContourPoints(cntPts)

                // Step 3: Crop the sudoku board out of the image and apply a perspective transform
                mRGBA = OpenCV.getCroppedSudokuBoard(mRGBA, sortedPoints)

                // Step 4: Convert the resulting Mat to an InputImage
                val bitmap = SudokuUtils.matToBitmap(mRGBA)
                val image = InputImage.fromBitmap(bitmap, 0)

                // Save the bitmap file to access it in the SolverActivity
                val fileOutputStream = baseContext.openFileOutput("tmpBitmap", Context.MODE_PRIVATE)
                bitmap.compress(Bitmap.CompressFormat.PNG, 100, fileOutputStream)
                fileOutputStream.close()

                // Step 5: Get the sudoku board (2d array) from the image using MLKit Text Recognition
                runBlocking {
                    val sudokuBoard = OCR.getSudokuFromImage(image)

                    // DEBUGGING
                    for (i in 0 until 9) {
                        val array = CharArray(9)
                        for (j in 0 until 9) {
                            array[j] = sudokuBoard[i][j].number.toString()[0]
                        }
                        Log.d(TAG, array.joinToString(" "))
                    }

                    // Step 6: Start the solving activity and finish this one
                    startActivity(Intent(baseContext, SolverActivity::class.java).apply {
                        putExtra("sudokuBoard", sudokuBoard)
                        addFlags(Intent.FLAG_ACTIVITY_NO_ANIMATION)
                    })
                }
            }
            STATUS = STATUS_PREVIEWING
        }
        return null
    }
}